File: 0206 - Reverse Linked List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        node1 = head
        if not head:
            return node1
        
        node2 = node1.next
        if not node2:
            return head
        
        node3 = node2.next
        if not node3:
            node1.next = None
            node2.next = node1
            return node2

        node1.next = None
        while node3:
            node2.next = node1
            node1 = node2
            node2 = node3
            node3 = node3.next
        
        node2.next = node1
        return node2

File: 0206 - Reverse Linked List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        vector<int> values;
        ListNode* other = head;

        while(other){
            values.push_back(other->val);
            other = other->next;
        }

        other = head;

        int index = values.size() - 1;
        while(other){
            other->val = values[index];
            other = other->next;
            index--;
        }

        return head;
    }
};